File: backend/app/models/puzzle.py
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from datetime import datetime

class PuzzlePiece(BaseModel):
    id: str
    image_path: str
    dimensions: Dict[str, float]  # x, y, z
    color_profile: List[int]  # RGB values
    shape_features: Dict[str, Any]
    position: Optional[Dict[str, float]] = None  # x, y, z coordinates in solution

class Puzzle(BaseModel):
    id: str
    name: str
    pieces: List[PuzzlePiece]
    target_shape: Optional[Dict[str, Any]] = None
    created_at: datetime
    solved: bool = False
    solution: Optional[List[Dict[str, Any]]] = None

class PuzzleUpload(BaseModel):
    name: str
    description: Optional[str] = None

class SolverRequest(BaseModel):
    puzzle_id: str
    algorithm: str = "genetic"  # genetic, simulated_annealing, reinforcement_learning
    max_iterations: int = 1000

class SolverResponse(BaseModel):
    puzzle_id: str
    solved: bool
    solution: Optional[List[Dict[str, Any]]]
    confidence: float
    processing_time: float

File: frontend/src/types/puzzle.ts
// Notch pattern for each side of a square piece
// Array of 6 booleans: true = present (material), false = notched (cut out)
export type NotchPattern = [boolean, boolean, boolean, boolean, boolean, boolean];

export interface SquarePieceNotches {
  top: NotchPattern;
  right: NotchPattern;
  bottom: NotchPattern;
  left: NotchPattern;
}

export interface PuzzlePiece {
  id: string;
  // For new square pieces with notches
  notches?: SquarePieceNotches;
  size?: number; // width/height of square (depth is always 1 notch unit)
  
  // Legacy fields for image-based pieces (keeping for backward compatibility)
  image_path?: string;
  dimensions?: {
    x: number;
    y: number;
    z: number;
  };
  color_profile?: number[];
  shape_features?: {
    area: number;
    perimeter: number;
    complexity: number;
    corners: number;
  };
  position?: {
    x: number;
    y: number;
    z: number;
  };
}

export interface Puzzle {
  id: string;
  name: string;
  pieces: PuzzlePiece[];
  target_shape?: any;
  created_at: string;
  solved: boolean;
  solution?: Array<{
    piece_id: string;
    position: {
      x: number;
      y: number;
      z: number;
    };
    rotation: {
      x: number;
      y: number;
      z: number;
    };
  }>;
}

export interface PuzzleUpload {
  name: string;
  description?: string;
}

export interface SolverRequest {
  puzzle_id: string;
  algorithm: 'genetic' | 'simulated_annealing' | 'reinforcement_learning';
  max_iterations: number;
}

export interface SolverResponse {
  puzzle_id: string;
  solved: boolean;
  solution?: Array<{
    piece_id: string;
    position: {
      x: number;
      y: number;
      z: number;
    };
    rotation: {
      x: number;
      y: number;
      z: number;
    };
  }>;
  confidence: number;
  processing_time: number;
}